addons/t9n/models/resource.py

```diff
@@ -14,3 +19,84 @@ class Resource(models.Model):
+            "unique(file_name, project_id)",
+            "A file with the same name already exists in the same project!"
+        ),
+    ]
+
+    def _decode_resource_file(self, resource_file):
+        try:
+            file_content = base64.b64decode(resource_file).decode()
+            po_obj = polib.pofile(file_content)
+        except (IOError, UnicodeDecodeError):
+            return None
+        return po_obj
+
+    def _get_message_from_file(self, resource_file):
+        for entry in resource_file:
+            references_list = [reference[0] + (f":{reference[1]}" if reference[1] else "")
+                               for reference in entry.occurrences]
+            entry.references = references_list
+
+        messages_to_create = [{"body": entry.msgid, "context": entry.msgctxt, "translator_comments": entry.tcomment,
+                               "extracted_comments": entry.comment, "references": "\n".join(entry.references)} for entry in resource_file]
+        return messages_to_create
+
+    def _get_messages_to_create(self, messages_in_db, messages_from_file):
+        return [Command.create(message) for message in messages_from_file if not any(message["body"] == db_message["body"] and message["context"] == db_message["context"] for db_message in messages_in_db)]
+
+    def _get_messages_to_unlink(self, messages_in_db, messages_from_file):
+        return [Command.unlink(db_message.id) for db_message in messages_in_db if not any(db_message["body"] == message["body"] and db_message["context"] == message["context"] for message in messages_from_file)]
+
+    def _get_messages_to_write(self, messages_in_db, messages_from_file):
+        return [Command.update(db_message.id, message) for db_message in messages_in_db for message in messages_from_file if db_message["body"] == message["body"] and db_message["context"] == message["context"]]
+
+    def _combine_messages_from_file(self, po_file_object):
+        messages_in_db = self.message_ids
+        messages_from_file = self._get_message_from_file(po_file_object)
+
+        messages_to_create = self._get_messages_to_create(messages_in_db, messages_from_file)
+        messages_to_unlink = self._get_messages_to_unlink(messages_in_db, messages_from_file)
+        messages_to_write = self._get_messages_to_write(messages_in_db, messages_from_file)
+
+        return messages_to_create + messages_to_unlink + messages_to_write
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        corrupted_files = []
+        successful_vals = []
+
+        for vals in vals_list:
+            if not vals.get("file"):
+                raise ValidationError(_("A resource file is required to create a resource."))
+            po_obj = self._decode_resource_file(vals["file"])
+            del vals["file"]
+
+            if po_obj is None:
+                corrupted_files.append(vals["file_name"])
+            else:
+                successful_vals.append({"vals": vals, "file": po_obj})
```

Comment:
I wouldn't call it file since the name is already used for something else. How about something like "po_data" instead?
```suggestion
                successful_vals.append({"vals": vals, "po_data": po_obj})
```